File: apps/api/src/db/migrate.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { pool } from "./pool.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const migrationsDir = path.resolve(__dirname, "../../migrations");

async function ensureMigrationsTable() {
  await pool.query(`
    CREATE TABLE IF NOT EXISTS schema_migrations (
      filename TEXT PRIMARY KEY,
      applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
    )
  `);
}

export async function migrate() {
  await ensureMigrationsTable();

  const files = (await fs.readdir(migrationsDir))
    .filter((file) => file.endsWith(".sql"))
    .sort();

  for (const filename of files) {
    const applied = await pool.query(
      "SELECT 1 FROM schema_migrations WHERE filename = $1",
      [filename],
    );

    if (applied.rowCount) {
      continue;
    }

    const sql = await fs.readFile(path.join(migrationsDir, filename), "utf8");
    await pool.query("BEGIN");

    try {
      await pool.query(sql);
      await pool.query("INSERT INTO schema_migrations (filename) VALUES ($1)", [
        filename,
      ]);
      await pool.query("COMMIT");
    } catch (error) {
      await pool.query("ROLLBACK");
      throw error;
    }
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  migrate()
    .then(() => {
      console.log("Migrations complete");
      process.exit(0);
    })
    .catch((error) => {
      console.error(error);
      process.exit(1);
    });
}

File: apps/api/app/db/migrate.py
import os
import psycopg2
from pathlib import Path
from app.core.config import settings

def run_migrations():
    """Run all pending SQL migrations"""
    migrations_dir = Path(__file__).parent.parent / "migrations"
    
    conn = psycopg2.connect(settings.DATABASE_URL)
    cursor = conn.cursor()
    
    # Create migrations tracking table
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS schema_migrations (
            version VARCHAR(255) PRIMARY KEY,
            applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        )
    """)
    conn.commit()
    
    # Get applied migrations
    cursor.execute("SELECT version FROM schema_migrations")
    applied = {row[0] for row in cursor.fetchall()}
    
    # Get migration files
    migration_files = sorted([
        f for f in os.listdir(migrations_dir) 
        if f.endswith('.sql') and f != 'schema.sql'
    ])
    
    # Run pending migrations
    for migration_file in migration_files:
        if migration_file in applied:
            print(f"✓ Already applied: {migration_file}")
            continue
        
        print(f"→ Applying: {migration_file}")
        
        migration_path = migrations_dir / migration_file
        with open(migration_path, 'r', encoding='utf-8') as f:
            sql = f.read()
        
        try:
            cursor.execute(sql)
            cursor.execute(
                "INSERT INTO schema_migrations (version) VALUES (%s)",
                (migration_file,)
            )
            conn.commit()
            print(f"✓ Applied: {migration_file}")
        except Exception as e:
            conn.rollback()
            print(f"✗ Failed: {migration_file}")
            print(f"  Error: {e}")
            break
    
    cursor.close()
    conn.close()
    print("\n✓ Migration complete")

if __name__ == "__main__":
    run_migrations()
